src/compiler/checker.ts

```diff
@@ -13573,23 +13597,30 @@ namespace ts {
          * This is used during inference when instantiating type parameter defaults.
          */
         function createBackreferenceMapper(context: InferenceContext, index: number): TypeMapper {
-            return t => findIndex(context.inferences, info => info.typeParameter === t) >= index ? unknownType : t;
+            return makeFunctionTypeMapper(t => findIndex(context.inferences, info => info.typeParameter === t) >= index ? unknownType : t);
         }
 
         function combineTypeMappers(mapper1: TypeMapper | undefined, mapper2: TypeMapper): TypeMapper;
         function combineTypeMappers(mapper1: TypeMapper, mapper2: TypeMapper | undefined): TypeMapper;
         function combineTypeMappers(mapper1: TypeMapper, mapper2: TypeMapper): TypeMapper {
-            if (!mapper1) return mapper2;
-            if (!mapper2) return mapper1;
-            return t => instantiateType(mapper1(t), mapper2);
+            return !mapper1 ? mapper2 : !mapper2 ? mapper1 : makeCompositeTypeMapper(mapper1, mapper2);
         }
 
-        function createReplacementMapper(source: Type, target: Type, baseMapper: TypeMapper): TypeMapper {
-            return t => t === source ? target : baseMapper(t);
+        function addTypeMapping(mapper: TypeMapper | undefined, source: TypeParameter, target: Type) {
+            return mapper && mapper.kind === TypeMapKind.Simple && mapper.source2 === mapper.target2 ?

```

Comment:
I'm getting mixed up.  When you combine two maps, you look in the first map first and then, if you find something, do you stop looking or apply the second map to the first target?  From the handling of composite maps in `getMappedType` it seems like it might be the latter, but this appears to do the former for unary maps?

Reply:
The particular test here is to see if the second source/target pair in the simple mapper is unused (recall, when source and target are the same, we have a no-op). If so, we create a new simple mapper with both source/target pairs in use.

One added twist with composite mappers is that the first mapper may map to some type that the second mapper further maps. For example, the first mapper might map from `T` to `U[]` and the second mapper from `U` to `string`. This also explains why we directly call the `getMappedType` with the first mapper, but then call `instantiateType` with the second one.

Reply:
I should add, in the `addTypeMapping` case, we actually don't care about the ability for the second mapping to affect the first mapping, which is why we can do the simple mapper optimization.

Reply:
So maps are combined by composition, but we happen to know that this particular composition is equivalent to concatenation?